File: tools/apigen/fidget.js
var ffi = require('ffi-napi');
var Struct = require("ref-struct-napi");

exports.cb = function(f){return ffi.Callback('void', [], f)};

var Node = Struct({'nimRef': 'pointer'});
EventCbKind = 'int64'
exports.E_ON_CLICK = 0
exports.E_ON_FRAME = 1
exports.E_ON_EDIT = 2
exports.E_ON_DISPLAY = 3
exports.E_ON_FOCUS = 4
exports.E_ON_UNFOCUS = 5


var dllPath = ""
if(process.platform == "win32") {
  dllPath = 'fidget.dll'
} else if (process.platform == "darwin") {
  dllPath = __dirname + '/libfidget.dylib'
} else {
  dllPath = __dirname + '/libfidget.so'
}

var fidget = ffi.Library(dllPath, {
  'fidget_on_click_global': ['void', ['pointer']],
  'fidget_node_get_name': ['string', [Node]],
  'fidget_node_set_name': ['void', [Node, 'string']],
  'fidget_node_get_characters': ['string', [Node]],
  'fidget_node_set_characters': ['void', [Node, 'string']],
  'fidget_node_get_dirty': ['bool', [Node]],
  'fidget_node_set_dirty': ['void', [Node, 'bool']],
  'fidget_find_node': [Node, ['string']],
  'fidget_add_cb': ['void', [EventCbKind, 'int64', 'string', 'pointer']],
  'fidget_start_fidget': ['void', ['string', 'string', 'string', 'bool']],
});

exports.onClickGlobal = fidget.fidget_on_click_global;
exports.Node = Node;
Object.defineProperty(Node.prototype, 'name', {
  get: function() {return fidget.fidget_node_get_name(this)},
  set: function(v) {fidget.fidget_node_set_name(this, v)}
});
Object.defineProperty(Node.prototype, 'characters', {
  get: function() {return fidget.fidget_node_get_characters(this)},
  set: function(v) {fidget.fidget_node_set_characters(this, v)}
});
Object.defineProperty(Node.prototype, 'dirty', {
  get: function() {return fidget.fidget_node_get_dirty(this)},
  set: function(v) {fidget.fidget_node_set_dirty(this, v)}
});
exports.findNode = fidget.fidget_find_node;
exports.addCb = fidget.fidget_add_cb;
exports.startFidget = fidget.fidget_start_fidget;

File: tools/apigen/fidget.rb
require 'ffi'

class Node < FFI::Struct
  layout ref: :uint64

  def name
    return DLL.fidget_node_get_name(self)
  end
  def name=(v)
    DLL.fidget_node_set_name(self, v)
  end

  def characters
    return DLL.fidget_node_get_characters(self)
  end
  def characters=(v)
    DLL.fidget_node_set_characters(self, v)
  end

  def dirty
    return DLL.fidget_node_get_dirty(self)
  end
  def dirty=(v)
    DLL.fidget_node_set_dirty(self, v)
  end
end

EventCbKind = :uint64
E_ON_CLICK = 0
E_ON_FRAME = 1
E_ON_EDIT = 2
E_ON_DISPLAY = 3
E_ON_FOCUS = 4
E_ON_UNFOCUS = 5


module DLL
  extend FFI::Library
  dll_path = ""
  if RUBY_PLATFORM == "x64-mingw32"
    dll_path = Dir.getwd + "/fidget.dll"
  elsif RUBY_PLATFORM == "arm64-darwin20"
    dll_path = Dir.getwd + "/libfidget.dylib"
  else
    dll_path = Dir.getwd + "/libfidget.so"
  end
  ffi_lib dll_path
  callback :fidget_cb, [], :void

  attach_function :fidget_on_click_global, [:fidget_cb], :void
  attach_function :fidget_node_get_name, [Node.by_value], :string
  attach_function :fidget_node_set_name, [Node.by_value, :string], :void
  attach_function :fidget_node_get_characters, [Node.by_value], :string
  attach_function :fidget_node_set_characters, [Node.by_value, :string], :void
  attach_function :fidget_node_get_dirty, [Node.by_value], :bool
  attach_function :fidget_node_set_dirty, [Node.by_value, :bool], :void
  attach_function :fidget_find_node, [:string], Node.by_value
  attach_function :fidget_add_cb, [EventCbKind, :int64, :string, :fidget_cb], :void
  attach_function :fidget_start_fidget, [:string, :string, :string, :bool], :void
end
def on_click_global(a)
  return DLL.fidget_on_click_global(a)
end

def find_node(glob)
  return DLL.fidget_find_node(glob)
end

def add_cb(kind, priority, glob, handler)
  return DLL.fidget_add_cb(kind, priority, glob, handler)
end

def start_fidget(figma_url, window_title, entry_frame, resizable)
  return DLL.fidget_start_fidget(figma_url, window_title, entry_frame, resizable)
end

